content/blog/2019-12-21-logical-indexing-on-HB/index.md

```diff
@@ -0,0 +1,249 @@
+
+
+# Specifying the Data Format
+By expressing the stencil algorithm in terms of logical indexing, we disentangle it from implementation details of data formatting and caching.
+
+We can specify these details by decorating inputs and outputs in the kernel signature as follows:
+```
+stencil_kernel(input[blocked cached, row-major] A : float[n][m], output[non-cached, row-major] B: float[n][m])
+```
+This signature specifies that both `A` and `B` are formatted in row-major order, that a block of `A` will be cached by each core, and that cores will write directly to `B` in global memory.
+Such a specification suggests to the compiler particular implementations for data formats $f_A, f_B$.
+
+# Stencil Algorithm with Logical Indexing in C
+Logical indexing and data formatting will be incorporated into my research language; however, compilation for this syntax has not been implemented yet. 
+Here, I share a prototype of logical indexing in C that generates cached and non-cached variants of the above stencil algorithm.
+
+
+We'll take the stencil algorithm as a diving board into the interplay of logical indexing and data formatting:
+```c
+// stencil algorithm
+void stencil(tensor A, tensor B) {
+  for (int x = 0; x < B.n; x++) {
+    for (int y = 0; y < B.m; y++) {
+      if (0 < log_to_phys_row(B, x) &&
+            log_to_phys_row(B, x) < top_level_tensor(B)-1 &&
+          0 < log_to_phys_col(B, y) &&
+            log_to_phys_col(B, y) < top_level_tensor(B)-1) {
+        log_write(B, x, y, avg(A, x, y));
+      }
+      else {
+        log_write(B, x, y, log_read(A, x, y));
+      }
+    }
+  }
+}
+```
+
+The translation from the earlier psuedo code is nearly straightforward: logical reads and writes are respectively denoted `log_read` and `log_write`.
+Note the `tensor` type, which collects data formatting information of an array to implement logical data operations; it's pretty [dope](https://en.wikipedia.org/wiki/Dope_vector):
+
+```c
+typedef struct tensor
+  { struct tensor *parent;
+    float* arr;
+    int n; int m;
+    int stride; 
+    // physical offset of logical origin in parent
+    int phys_row_off; int phys_col_off; 
+    bool cached;
+  } tensor;
+```
+
+Using this information, we implement `log_read` as follows:
+```c
+  float log_read(tensor *A, int log_row, int log_col) {
+    int phys_row = log_to_phys_row(A, log_row);
+    int phys_col = log_to_phys_col(A, log_col);
+
+    return A->arr[phys_row * A->stride + phys_col];
+  }
+```
+
+`log_to_phys_row` applies sub-tensor offsets to the logical row index recursively up the nested sub-tensor structure:
+```c
+int log_to_phys_row(tensor *A, int log_row) {
+  return A->cached || A->parent == NULL ? 
+    log_row + A->phys_row_off:
+    log_to_phys_row(A->parent, log_row + A->phys_row_off);
+}
+```
```

Comment:
Seems cool! It does occur to me that this _literal_ code you are listing here, with the extra metadata, may be overkill for code generation. If you know the size, stride, etc. _statically_, then it might make sense to inline those values here statically rather than looking them up dynamically.

Reply:
Good point. Yes, this is work that I envision the compiler doing statically during code generation rather than dynamically. Since I embedded logical indexing directly in C, the simpler alternative for this project was just to implement it's behavior dynamically as opposed to doing fancy pre-processing / code-gen.

I made a note of this where I define these expensive lookup functions.